Prevent Hero transition for AppBar on Pesto recipe page.

mpcomplete opened the pull request:
BUG=https://github.com/flutter/flutter/issues/5351

mpcomplete committed:
Prevent Hero transition for AppBar on Pesto recipe page.

BUG=https://github.com/flutter/flutter/issues/5351

abarth commented:
Would this look better if we cross-faded from the transparent app bar on the recipe page to the green one on the list page?

abarth commented on examples/flutter_gallery/lib/demo/pesto_demo.dart:
```diff
@@ -327,6 +327,7 @@ class _RecipePageState extends State<RecipePage> {
       scrollableKey: _scrollableKey,
       appBarBehavior: AppBarBehavior.scroll,
       appBar: new AppBar(
+        heroTag: 'nohero',
```
Should we have a distinguished value (e.g., `null`) rather than using this random string, which might match some other hero tag that also uses that random string?

HansMuller commented:
LGTM

I agree with Adam about providing a conventional way to opt out of the AppBar hero transition.  Maybe file an issue about that.

Hixie commented:
We should have the distinguished way in this PR, imho

mpcomplete committed:
Merge branch 'pesto.appbar' of github.com.:mpcomplete/flutter into 0

mpcomplete committed:
null.heroTag

mpcomplete committed:
Revert "null.heroTag"

This reverts commit 50f7be51c13bb449f7bd945be910de90f246cac8.

abarth commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -14,7 +14,8 @@ import 'tabs.dart';
 import 'typography.dart';
 
-final Object _kDefaultHeroTag = new Object();
+const Object _kDefaultHeroTag = const Object();
```
We can't use `const Object()` as the default hero tag because const objects are de-duplicated across the system.  That means another file that makes a `const Object()` hero tag will match this one.

mpcomplete commented:
> Would this look better if we cross-faded from the transparent app bar on the recipe page to the green one on the list page?

I'm undecided on whether this looks better (I tried this by removing the `expandedHeight` arg to the recipe page's AppBar). But I'm also not sure how to do this while also keeping the same scroll behavior. With that arg present, scrolling the recipe page doesn't move the app bar until the text area meets the app bar. Without the arg, scrolling immediately moves the app bar.

> Should we have a distinguished value (e.g., null) rather than using this random string, which might match some other hero tag that also uses that random string?

I tried this as well. See https://github.com/flutter/flutter/pull/5760/commits/50f7be51c13bb449f7bd945be910de90f246cac8 . I reverted that, though, and replaced the 'nohero' string with a new Object. IMO that's clearer than a special value, and matches the way Keys work.

mpcomplete committed:
custom.obj

Hixie commented:
somehow your patch ended up containing another PR's changes? I'm confused.

mpcomplete committed:
Merge remote-tracking branch 'upstream/master' into pesto.appbar

mpcomplete commented:
Weird - new upload fixes it.

Hixie commented:
LGTM
